crates/ruff_python_formatter/src/expression/string.rs

```diff
@@ -1160,13 +1299,99 @@ impl<'ast, 'buf, 'fmt> DocstringLinePrinter<'ast, 'buf, 'fmt> {
         // We handled the case that the closing quotes are on their own line
         // above (the last line is empty except for whitespace). If they are on
         // the same line as content, we don't insert a line break.
-        if !self.is_last {
+        if !line.is_last {
             hard_line_break().fmt(self.f)?;
         }
 
         Ok(())
     }
 
+    /// Given a sequence of lines from a code snippet, format them and return
+    /// the formatted code as a sequence of owned docstring lines.
+    ///
+    /// This routine generally only returns an error when the recursive call
+    /// to the formatter itself returns a `FormatError`. In all other cases
+    /// (for example, if the code snippet is invalid Python or even if the
+    /// resulting reformatted code snippet is invalid Python), then `Ok(None)`
+    /// is returned. In this case, callers should assume that a reformatted
+    /// code snippet is unavailable and bail out of trying to format it.
+    ///
+    /// Currently, when the above cases happen and `Ok(None)` is returned, the
+    /// routine is silent about it. So from the user's perspective, this will
+    /// fail silently. Ideally, this would at least emit a warning message,
+    /// but at time of writing, it wasn't clear to me how to best do that.
+    fn format(
+        &mut self,
+        code: &[CodeExampleLine<'src>],
+    ) -> FormatResult<Option<Vec<DocstringLine<'static>>>> {
+        use ruff_python_parser::AsMode;
+
+        let offset = code
+            .get(0)
+            .expect("code blob must be non-empty")
+            .original
+            .offset;
+        let last_line_is_last = code
+            .last()
+            .expect("code blob must be non-empty")
+            .original
+            .is_last;
+        let codeblob = code
+            .iter()
+            .map(|line| &*line.code)
+            .collect::<Vec<&str>>()
+            .join("\n");
+        let printed = match docstring_format_source(self.f.options(), self.quote_style, &codeblob) {
+            Ok(printed) => printed,
+            Err(FormatModuleError::FormatError(err)) => return Err(err),
+            Err(
+                FormatModuleError::LexError(_)
+                | FormatModuleError::ParseError(_)
+                | FormatModuleError::PrintError(_),
+            ) => {
+                return Ok(None);
+            }
+        };
+        // This is a little hokey, but we want to determine whether the
```

Comment:
We should create an issue for removing this, parsing twice every time seems wasteful